arch/arm/src/gd32f4/gd32f4xx_sdio.c

```diff
@@ -0,0 +1,3070 @@
+  priv->xfrflags   = 0;
+#endif
+#ifdef CONFIG_GD32_SDIO_CARD
+  putreg32(priv->xfrmask | priv->waitmask | priv->sdiointmask,
+           GD32_SDIO_INTEN);
+
+#else
+  putreg32(priv->xfrmask | priv->waitmask, GD32_SDIO_INTEN);
+#endif
+
+  leave_critical_section(flags);
+}
+
+/****************************************************************************
+ * Name: gd32_config_xfrints
+ *
+ * Description:
+ *   Enable SDIO interrupts needed to support the data transfer event
+ *
+ * Input Parameters:
+ *   priv    - A reference to the SDIO device state structure
+ *   xfrmask - The set of bits in the SDIO INTEN register to set
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void gd32_config_xfrints(struct gd32_dev_s *priv, uint32_t xfrmask)
+{
+  irqstate_t flags;
+
+  flags = enter_critical_section();
+  priv->xfrmask = xfrmask;
+#ifdef CONFIG_GD32_SDIO_CARD
+  putreg32(priv->xfrmask | priv->waitmask | priv->sdiointmask,
+           GD32_SDIO_INTEN);
+#else
+
+  putreg32(priv->xfrmask | priv->waitmask, GD32_SDIO_INTEN);
+#endif
+  leave_critical_section(flags);
+}
+
+/****************************************************************************
+ * Name: gd32_set_pwrctl
+ *
+ * Description:
+ *   Change the PWRCTRL field of the SDIO POWER register to turn the SDIO
+ *   ON or OFF
+ *
+ * Input Parameters:
+ *   clkctl - A new PWRCTRL setting
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void gd32_set_pwrctl(uint32_t pwrctl)
+{
+  uint32_t regval;
+
+  regval  = getreg32(GD32_SDIO_PWRCTL);
+  regval &= ~SDIO_PWRCTL_MASK;
+  regval |= pwrctl;
+  putreg32(regval, GD32_SDIO_PWRCTL);
+}
+
+/****************************************************************************
+ * Name: gd32_sampleinit
+ *
+ * Description:
+ *   Setup prior to collecting DMA samples
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SDIO_XFRDEBUG
+static void gd32_sampleinit(void)
+{
+  memset(g_sampleregs, 0xff,
+         DEBUG_NSAMPLES * sizeof(struct gd32_sampleregs_s));
+}
+#endif
+
+/****************************************************************************
+ * Name: gd32_sdiosample
+ *
+ * Description:
+ *   Sample SDIO registers
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SDIO_XFRDEBUG
+static void gd32_sdiosample(struct gd32_sdioregs_s *regs)
+{
+  regs->power   = (uint8_t)getreg32(GD32_SDIO_PWRCTL);
+  regs->clkctl   = (uint16_t)getreg32(GD32_SDIO_CLKCTL);
+  regs->datactl   = (uint16_t)getreg32(GD32_SDIO_DATACTL);
+  regs->datato  = getreg32(GD32_SDIO_DATATO);
+  regs->datalen    = getreg32(GD32_SDIO_DATALEN);
+  regs->datacnt  = getreg32(GD32_SDIO_DATACNT);
+  regs->stat     = getreg32(GD32_SDIO_STAT);
+  regs->inten    = getreg32(GD32_SDIO_INTEN);
+  regs->fifocnt = getreg32(GD32_SDIO_FIFOCNT);
```

Comment:
Please align to all "=" be at same position, see stm32_sdiosample() for reference